pandas/core/generic.py

```diff
@@ -11817,36 +11818,70 @@ def pct_change(
         GOOG  0.179241  0.094112   NaN
         APPL -0.252395 -0.011860   NaN
         """
-        # GH#53491
-        if fill_method is not lib.no_default or limit is not lib.no_default:
+        # GH#53491: deprecate the `fill_method` and `limit` keyword, except
+        # `fill_method=None` that does not fill missing values
+        if fill_method not in (lib.no_default, None) and limit is not lib.no_default:
```

Comment:
I don't think we need warning messages for every single case. Does something like

> The 'fill_method' being not None and the 'limit' argument are deprecated. Either fill in NA values prior to calling pct_change or specify fill_method=None.

work?

Reply:
I think we may want to prompt users to use `fill_method=None` even if they have filled NA values. See the original comment of this PR. The case is, even after `ffill` or `bfill`, calling `pct_change` without keyword will raise deprecation warning unless we explicitly check if there are NA values to fill. However, I don't think this is a good approach: (1) this may add too much overhead, and (2) if a user is not filling NA values and uses `pct_change` without keyword, and if the data occasionally does not contain NA values, he/she will not get a warning message and the logic would be incorrect.

Due to these reasons, I think this deprecation would be especially confusing, especially since we are having "incorrect" deprecation warnings in the current version. That's why I'm trying to give extra specific guide for each case. If maintainers do not think this is necessary, I can implement using only a single message.

Reply:
Still, I need a confirmation about whether we should prompt users to do `obj.ffill/bfill().pct_change(fill_method=None)` or `obj.ffill/bfill().pct_change()`. I prefer the former.

Reply:
> However, I don't think this is a good approach: (1) this may add too much overhead, and (2) if a user is not filling NA values and uses `pct_change` without keyword, and if the data occasionally does not contain NA values, he/she will not get a warning message and the logic would be incorrect.

I'm seeing the current logic takes 7.5% of the runtime for the current warning on a Series with 100k rows - I don't think overhead is a concern. This will cause users to modify their code unnecessarily in what I think is the uncommon case. They will then need to change their code again when we deprecate the `fill_method` argument. I do not think we should do that.

> The thing is, even if users do obj.ffill/bfill().pct_change(), a warning will still be raised unless we check for NA values

That is why I am suggesting the language "Either fill in NA values prior to calling..." and not saying to use ffill/bfill.

Reply:
Sure, I will implement your suggestions.

Reply:
Perhaps "Either fill in any non-leading NA values" is better.

Reply:
agreed with @rhshadrach